### PR TITLE
Implement SciPy-compatible FFT functions

### PR DESCRIPTION
**Pull request by not522**
This PR implements SciPy-compatible FFT functions. They use in-place FFT if `overwrite_x` is `True` and transforms are C2C, which is requested at #1669.

cuFFT also supports in-place R2C and C2R FFTs, but they are not recommended (see https://docs.nvidia.com/cuda/cufft/index.html#accuracy-and-performance), therefore we use out-of-place mode in such cases.

### CODE REVIEW

**Comment by takagi** [comment]
@toslunar Can I take over reviewing this PR?

**Comment by toslunar** [comment]
@takagi Sure.  Thanks.

File: docs/source/_comparison_generator.py [hunk 1]
```diff
@@ -107,6 +107,9 @@ def generate():
     buf += _section(
         'Sparse Matrices',
         'scipy.sparse', 'cupyx.scipy.sparse', 'SciPy')
+    buf += _section(
```
Comment (takagi) [inline docs/source/_comparison_generator.py]: Would you move this section before "Sparse Matrices" to follow SciPy's API reference list if you don't have a special reason?

https://docs.scipy.org/doc/scipy/reference/index.html#api-reference

**Comment by takagi** [comment]
LGTM except the comments above!

**Comment by not522** [comment]
I fixed documents according to your comment. Also, test fails were fixed.

**Comment by takagi** [comment]
Jenkins, test this please.

**Comment by takagi** [comment]
LGTM, will merge after CI finish!

**Pull request merged**